aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,408 @@
+    a query (currently UpdateRule and ReplaceRule).
+    """
+
+    def __init__(self, querybuilder, max_iterations=1, track_edges=False):
+        """
+        :param querybuilder: An instance of the QueryBuilder class
+        :param int max_iterations: The number of iterations to run this query on
+        :param bool track_edges: Whether to track which edges are traversed and store them
+        """
+
+        def get_spec_from_path(queryhelp, idx):
+            if (
+                queryhelp['path'][idx]['entity_type'].startswith('node') or
+                queryhelp['path'][idx]['entity_type'].startswith('data') or
+                queryhelp['path'][idx]['entity_type'].startswith('process') or
+                queryhelp['path'][idx]['entity_type'] == ''
+            ):
+                # not really nice, this way, how about using
+                # from aiida.orm.utils.node import is_valid_node_type_string and checking entity_type?
+                result = 'nodes'
+            elif queryhelp['path'][idx]['entity_type'] == 'group':
+                result = 'groups'
+            else:
+                raise Exception('not understood entity from ( {} )'.format(queryhelp['path'][idx]['entity_type']))
+            return result
+
+        queryhelp = querybuilder.queryhelp
+
+        # Check if there is any projection:
+        query_projections = queryhelp['project']
+        for projection_key in query_projections:
+            if query_projections[projection_key] != []:
+                error_message = 'The input querybuilder must not have any projections in it.\n'
+                error_message = error_message + 'Instead, it has the following:\n'
+                error_message = error_message + ' - Key: {}\n'.format(projection_key)
+                error_message = error_message + ' - Val: {}\n'.format(query_projections[projection_key])
+                raise ValueError(error_message)
+
+        for pathspec in queryhelp['path']:
+            if not pathspec['entity_type']:
+                pathspec['entity_type'] = 'node.Node.'
+        self._qbtemplate = QueryBuilder(**queryhelp)
+        queryhelp = self._qbtemplate.queryhelp
+        self._first_tag = queryhelp['path'][0]['tag']
+        self._last_tag = queryhelp['path'][-1]['tag']
+        self._querybuilder = None
+
+        # All of these are set in _init_run:
+        self._edge_label = None
+        self._edge_keys = None
+        self._entity_to_identifier = None
+
+        self._entity_from = get_spec_from_path(queryhelp, 0)
+        self._entity_to = get_spec_from_path(queryhelp, -1)
+        self._accumulator_set = None
+        super().__init__(max_iterations, track_edges=track_edges)
+
+    def set_edge_keys(self, edge_keys):
+        """
+        Set the edge keys that are use to classify the edges during the run of this query.
+        :param edge_keys: A list of projections on the edge itself, or a tuple that
```

Comment:
do the docs pass?
from my experience you need to indent the continued explanation

Reply:
Ok, I see; yes, the docs pass, but they are probably rendering garbage. Is it also mandatory to add an empty line between the general description and the params?